src/sys/socket/addr.rs

```diff
@@ -515,15 +515,44 @@ impl fmt::Display for Ipv6Addr {
 }
 
 /// A wrapper around `sockaddr_un`.
-///
-/// This also tracks the length of `sun_path` address (excluding
-/// a terminating null), because it may not be null-terminated.  For example,
-/// unconnected and Linux abstract sockets are never null-terminated, and POSIX
-/// does not require that `sun_len` include the terminating null even for normal
-/// sockets.  Note that the actual sockaddr length is greater by
-/// `offset_of!(libc::sockaddr_un, sun_path)`
 #[derive(Clone, Copy, Debug)]
-pub struct UnixAddr(pub libc::sockaddr_un, pub usize);
+pub struct UnixAddr {
+    // INVARIANT: sun+sun_len is valid as defined by docs for from_raw_parts
+    sun: libc::sockaddr_un,
+    sun_len: usize
+}
+
+// linux man page unix(7) says there are 3 kinds of unix socket:
+// pathname: addrlen = offsetof(struct sockaddr_un, sun_path) + strlen(sun_path) + 1
+// unnamed: addrlen = sizeof(sa_family_t)
+// abstract: addren > sizeof(sa_family_t), name = sun_path[..(addrlen - sizeof(sa_family_t))]
+//
+// what we call "len"/"sun_len" = addrlen - offsetof(struct sockaddr_un, sun_path)
+#[derive(PartialEq, Eq, Hash)]
+enum UnixAddrKind<'a> {
+    Pathname(&'a Path),
+    Unnamed,
+    #[cfg(any(target_os = "android", target_os = "linux"))]
+    Abstract(&'a [u8]),
+}
+impl<'a> UnixAddrKind<'a> {
+    /// Safety: sockaddr+size must be valid
+    #[allow(unused_unsafe)] // #[deny(unsafe_op_in_unsafe_fn)]
```

Comment:
This `allow(unused_unsafe)` is a red flag.  Why did you make the function unsafe?  Either the function itself should be unsafe, or there should be unsafe blocks around `slice::from_raw_parts`, but not both.

Reply:
I was sort of following [RFC 2585](https://github.com/rust-lang/rfcs/blob/master/text/2585-unsafe-block-in-unsafe-fn.md) (more accidentally than anything), but it would probably just be easier if I just get rid of the unsafe blocks.

Reply:
Ahh, I wasn't aware of that RFC.  Is it implemented yet?

Reply:
Yes, and I did use the uncommented `unsafe_op_in_unsafe_fn` but it was stabilized fairly recently so it didn't pass CI.